rotkehlchen/chain/ethereum/modules/curve/decoder.py

```diff
@@ -257,35 +259,155 @@ def _decode_curve_events(  # pylint: disable=no-self-use
-        to_address = hex_or_bytes_to_address(tx_log.topics[2])
-        if (  # deposit give asset
-            event.event_type == HistoryEventType.RECEIVE and
-            event.event_subtype == HistoryEventSubType.NONE and
-            source_address == CURVE_Y_DEPOSIT and
-            transaction.from_address == to_address
-        ):
-            crypto_asset = event.asset.resolve_to_crypto_asset()
-            event.event_type = HistoryEventType.WITHDRAWAL
-            event.event_subtype = HistoryEventSubType.REMOVE_ASSET
-            event.counterparty = CPT_CURVE
-            event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} from the curve pool {CURVE_Y_DEPOSIT}'  # noqa: E501
-            return True
-        return False
+        receive_event_idx, deposit_events = None, []
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.DEPOSIT and
+                event.event_subtype == HistoryEventSubType.DEPOSIT_ASSET and
+                event.counterparty == CPT_CURVE
+            ):
+                deposit_events.append(event)
+
+        # exit early if previously decoded deposit events aren't present.
+        if len(deposit_events) == 0:
+            log.debug(
+                f'Expected to see a receive pool token event and deposit events for a curve pool, '
+                f'but have not found them. Tx_hash: {transaction.tx_hash.hex()}',
+            )
+            return decoded_events
+
+        # search for the receipt of the wrapped token after deposit
+        for idx, event in enumerate(decoded_events):
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == transaction.from_address and
+                event.counterparty in CURVE_DEPOSIT_CONTRACTS
+            ):
+                receive_event_idx = idx
+                crypto_asset = event.asset.resolve_to_crypto_asset()
+                event.event_type = HistoryEventType.RECEIVE
+                event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                event.counterparty = CPT_CURVE
+                event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} after depositing in curve pool'  # noqa: E501
+                break
+
+        # exit early if there's no event for the wrapped token receipt.
+        if receive_event_idx is None:
+            log.debug(
+                f'Expected to see a receive pool token event and deposit events for a curve pool, '
+                f'but have not found them. Tx_hash: {transaction.tx_hash.hex()}',
+            )
+            return decoded_events
+
+        decoded_events[receive_event_idx].extra_data = {'deposit_events_num': len(deposit_events)}  # for accounting  # noqa: E501
+        previous_event = decoded_events[receive_event_idx]
+        for event in deposit_events:
```

Comment:
If I am not mistaken then the logic from `_handle_zap_contracts_liquidity_addition`, `_handle_zap_contracts_liquidity_removal`, `_decode_curve_deposit_events` and `_decode_curve_remove_events` can be abstracted.

Reply:
The only piece of code that is common to them is
```python
        decoded_events[return_event_idx].extra_data = {'withdrawal_events_num': len(withdrawal_events)}  # for accounting  # noqa: E501
        previous_event = decoded_events[return_event_idx]
        for event in withdrawal_events:
            maybe_reshuffle_events(previous_event, event, decoded_events)
            previous_event = event
```

Or are you talking about something else?

Reply:
Yeah, at least this. Also in general labeling deposit / withdraw events and receive / return wrapped tokens functionality looks same to me.

Reply:
> Also in general labeling deposit / withdraw events and receive / return wrapped tokens functionality looks same to me

Yeah to an extent. But combining/abstracting in this case will result in it being harder to read but the mention I pointed out above has been abstracted.

If you feel that's not the case, for sure I'd abstract the said logics.